### PR TITLE
Save and load recipe books for mass.off

### PR DESCRIPTION
**Pull request by ggggggggg**
This PR gets saving and loading of recipe books working for mass.off. It makes two changes to support this, only one of which actually matters at the moment.
1. replaces many usages of lambda with custom defined classes that live in `recipe_classes.py`
2. Change the way `coef_dtype` is used to allow indexing with both `filtValue` and `coefs`

Regarding 1. In the past, I used `dill` instead of `pickle` to enable saving a labmdas. At some point this stopped working. Now it seems to work fine again. The saving and loading lambdas is listed as one of the "major features" of dill, so it sure seems like they intend it to work. Removing the lambdas makes the code a bit harder to read, but otherwise has little effect. I'm tempted to leave in this work even though it currently seems unnecessary. In case dill breaks again, we'll be halfway to working without dill. Currently there are still a few usages of lambdas in tests, but not in the main codebase.

The second change has to do with a feature I probably shouldn't have added, which is like a view in a database. Basically the same column of data can be indexed a few different ways. The way I did it before, the `Recipe` has a reference to an `OffFile` and therefore to a `mmap`. Now `RecipeBook` only stores a `numpy.dtype`, which is easily pickleable.

I added a new test file that uses 5lag filters with saving and loading of recipe books.

### CODE REVIEW

**Comment by joefowler** [comment]
Do these two warnings during tests worry you? This is what I see locally, but the GitHub Actions page shows the same.
```text
===================================================================== warnings summary =====================================================================
tests/off/test_5lag_off.py::test_off_5lag_with_saving_and_loading_recipes
  /Users/fowlerj/qsp/lib/python3.11/site-packages/numpy/core/fromnumeric.py:3504: RuntimeWarning: Mean of empty slice.
    return _methods._mean(a, axis=axis, dtype=dtype,

tests/off/test_5lag_off.py::test_off_5lag_with_saving_and_loading_recipes
  /Users/fowlerj/qsp/lib/python3.11/site-packages/numpy/core/_methods.py:129: RuntimeWarning: invalid value encountered in divide
    ret = ret.dtype.type(ret / rcount)

-- Docs: https://docs.pytest.org/en/stable/how-to/capture-warnings.html
======================================================= 188 passed, 1 xfailed, 2 warnings in 35.80s ========================================================
```

**Comment by joefowler** [comment]
Don't think I 100% understand your first point. Do you mean that these changes removed _most_ but not all uses of `dill`? That would be nice. But if they remove _all_ uses, then let's dump that dependency. It's been a pain in the past.

File: tests/off/.gitignore [hunk 1]
```diff
@@ -0,0 +1 @@
+test_recipe_book_save.pkl
```
Comment (joefowler) [inline tests/off/.gitignore:1]: Fix missing newline.

File: tests/off/test_channels.py [hunk 2]
```diff
@@ -477,6 +477,15 @@ def funb(a, z):
...
     assert rb.craft(lambda a, b, c: a + b + c, args) == 18
 
+    rb.to_file("test_recipe_book_save.pkl", overwrite=True)
+
+    rb2 = util.RecipeBook.from_file("test_recipe_book_save.pkl")
+
+    assert rb2.craft("a", args) == 3
+    assert rb2.craft("b", args) == 6
+    assert rb2.craft("c", args) == 9
+    assert rb2.craft("c", args) == 9
```
Comment (joefowler) [inline tests/off/test_channels.py:487]: Does it make sense to have this line 2x?

**Review by joefowler — CHANGES_REQUESTED** [review]
You should fix the duplicated test and the missing newline.
Otherwise, I don't really know what I'm looking at, but I'm glad to see two tests move out of the expected-to-fail state!

**Review by joefowler — CHANGES_REQUESTED** [review]
Also, running the tests cluttered my main MASS directory with several files:
```
(qsp) mass ➤ ll -t | awk '{print $1, $6, $7, $8, $9}'                                                                                git:save_load_recipes*
total    
-rw-r--r-- May 14 12:51 20181205_BCDEFGHI_hists.hdf5
-rw-r--r-- May 14 12:51 20181205_BCDEFGHI_listmode.hdf5
-rw-r--r-- May 14 12:51 test_recipe_book_save.pkl
-rw-r--r-- May 14 12:51 recipe_book_save_test2.rbpkl
-rw-r--r-- May 14 12:50 20181205_BCDEFGHI_qualitychecklinefit.hdf5
```

Seems like these belong in a temporary directory, for when we run tests locally.

File: tests/off/.gitignore (hunk 1 above)
Comment (joefowler) [reply tests/off/.gitignore:1]: Actually, I take that back. We should _not_ be ignoring these pickle files. We should be placing them somewhere else!

**Comment by joefowler** [comment]
I made some updates, saving recipe pickles in temporary files. Seem reasonable?

**Comment by ggggggggg** [comment]
> Don't think I 100% understand your first point. Do you mean that these changes removed _most_ but not all uses of `dill`? That would be nice. But if they remove _all_ uses, then let's dump that dependency. It's been a pain in the past.

Currently the main code (in mass/off) doesn't need dill, but some of the tests do need dill because they use lambdas. So if dill fails on us again, I think it will be easy to fix.

**Review by joefowler — APPROVED** [review]

**Pull request merged**